### PR TITLE
EVG-8026 connect struct and conversion code generation

### PR DESCRIPTION
**Pull request by john-m-liu**
This connects the 2 previous parts. Given a schema for what you want the REST model to look like as well as the db model it's connected to, we will generate the struct and conversion code. Still needs work on handling complex types and validation (coming soon)

### CODE REVIEW

File: rest/model/codegen.go [hunk 1]
```diff
@@ -233,11 +252,18 @@ func generateServiceConversions(structVal *types.Struct, packageName, structName
...
+	fieldErrs := grip.NewBasicCatcher()
 	for i := 0; i < structVal.NumFields(); i++ {
 		field := structVal.Field(i)
 		fieldName := field.Name()
 		if fieldInfo, shouldExtract := fields[fieldName]; shouldExtract {
 			// generate the BuildFromService code
+			if !strings.Contains(fieldInfo.OutputFieldType, field.Type().String()) && !strings.Contains(field.Type().String(), fieldInfo.OutputFieldType) {
+				// this should ideally be a more sophisticated check to ensure that complex types are convertible
+				// to each other, but for now we rely on the naming convention to find obvious type errors
+				fieldErrs.Errorf("DB model field '%s' has type %s which is incompatible with REST model type %s", fieldName, field.Type().String(), fieldInfo.OutputFieldType)
```
Comment (ablack12) [inline rest/model/codegen.go:264]: Can you put '%s' for both types?

File: rest/model/codegen.go [hunk 2]
```diff
@@ -64,7 +60,10 @@ type modelConversionInfo struct {
 	TsConversions  string
 }
 
-func SchemaToGo(schema string) ([]byte, error) {
+// ModelMapping maps schema type names to their respective DB model
+type ModelMapping map[string]string
+
+func Codegen(schema string, config ModelMapping) ([]byte, error) {
```
Comment (ablack12) [inline rest/model/codegen.go:66]: Is it correct that this isn't used anywhere yet?

File: rest/model/codegen.go [hunk 3]
```diff
@@ -233,11 +252,18 @@ func generateServiceConversions(structVal *types.Struct, packageName, structName
 	}
 	bfsCode := []string{}
 	tsCode := []string{}
+	fieldErrs := grip.NewBasicCatcher()
 	for i := 0; i < structVal.NumFields(); i++ {
 		field := structVal.Field(i)
 		fieldName := field.Name()
 		if fieldInfo, shouldExtract := fields[fieldName]; shouldExtract {
 			// generate the BuildFromService code
+			if !strings.Contains(fieldInfo.OutputFieldType, field.Type().String()) && !strings.Contains(field.Type().String(), fieldInfo.OutputFieldType) {
```
[the remark this answers is not in the document]
Comment (john-m-liu) [reply rest/model/codegen.go:261]: done

File: rest/model/codegen.go (hunk 1 above)
Comment (john-m-liu) [reply rest/model/codegen.go:264]: done

File: rest/model/testdata/schema/basic.graphql [hunk 4]
```diff
@@ -1,13 +1,4 @@
-type MyType {
-  stringField: String
-  AnotherStringField: String
-  aTime: Time
-  number: Int
-  complex: [What!]
```
[the remark this answers is not in the document]
Comment (john-m-liu) [reply rest/model/testdata/schema/basic.graphql:6]: The test now requires actual go structs, so I'm using ones that we have in our model package

**Pull request merged**